File: takehome/app/src/main/java/com/fivestars/communication/CommunicationPlugin.java
package com.fivestars.communication;


import android.content.Intent;

import com.fivestars.takehome.ChatHeadService;
import com.fivestars.takehome.MainActivity;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CordovaWebView;
import org.json.JSONArray;
import org.json.JSONException;

/**
 * This class echoes a string called from JavaScript.
 */
public class CommunicationPlugin extends CordovaPlugin {

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("coolMethod")) {
            String message = args.getString(0);
            this.coolMethod(message, callbackContext);
            return true;
        }
        return false;
    }

    private void coolMethod(String message, CallbackContext callbackContext) {
        if (message != null && message.length() > 0) {
            startChatHeadService();
            callbackContext.success(message);
        } else {
            callbackContext.error("Expected one non-empty string argument.");
        }
    }

    /**
     * Set and initialize the view elements.
     */
    private void startChatHeadService() {
        cordova.getContext().startService(new Intent(cordova.getContext(), ChatHeadService.class));
    }

}

File: takehome/app/src/main/java/com/fivestars/communication/CommunicationPlugin.kt
package com.fivestars.communication


import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.json.JSONArray
import org.json.JSONException


/**
 * This class echoes a string called from JavaScript.
 */
class CommunicationPlugin : CordovaPlugin() {

    private var communicationListener: CommunicationPlugin.CommunicationListener? = null


    @Throws(JSONException::class)
    override fun execute(action: String, args: JSONArray, callbackContext: CallbackContext): Boolean {
        communicationListener?.let {
            it.execute(action, callbackContext)
        }
        return true
    }

    fun setCommunicationListener(communicationListener: CommunicationListener) {
        this.communicationListener = communicationListener
    }

    interface CommunicationListener {
        fun execute(action: String, callbackContext: CallbackContext)
    }
}
